### PR TITLE
OGM-1483 Define Infinispan Remote Protobuf schema providing the final file

### PR DESCRIPTION
**Pull request by fax4ever**
https://hibernate.atlassian.net/browse/OGM-1483

### CODE REVIEW

File: infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/protobuf/SchemaDefinitions.java [hunk 1]
```diff
@@ -64,7 +75,44 @@ public void deploySchema(String generatedProtobufName, RemoteCache<String, Strin
...
+
+	private void checkCompatibility(String generated, String providedByService) {
+		String providedByServiceWithoutComments = providedByService.trim()
+				.replaceAll( "\\/\\*([\\S\\s]+?)\\*\\/", "" )
+				.replaceAll( "(?s)/\\*.*?\\*/", "" );
+
+		String generatedWithoutSpaces = generated.replaceAll( "\\s+", " " ).trim();
+		String providedByServiceWithoutCommentsAndSpaces = providedByServiceWithoutComments.replaceAll( "\\s+", " " ).trim();
+
+		if ( !generatedWithoutSpaces.equalsIgnoreCase( providedByServiceWithoutCommentsAndSpaces ) ) {
```
Comment (fax4ever) [inline infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/protobuf/SchemaDefinitions.java]: @Sanne I don't like very much this solution... I'm going to figured out something better exploring Infinispan code.

Comment (Sanne) [reply infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/protobuf/SchemaDefinitions.java]: Cool. I'll stop the review then. Let me know when it's done.

File: infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/schema/TableDefinition.java [hunk 2]
```diff
@@ -117,4 +119,33 @@ public void validate() {
 	public String getCacheConfiguration() {
 		return cacheConfiguration;
 	}
+
+	public boolean isDescribedIn(FileDescriptor fileDescriptor) {
```
Comment (fax4ever) [inline infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/schema/TableDefinition.java]: All these **isDescribedIn** methods could have implemented with a visitor pattern, but I've preferred to using the simple OO delegation. WDYT @Sanne?

Comment (DavideD) [reply infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/schema/TableDefinition.java]: The pattern that `Shall Not Be Named`! :D 

The problem with the visitor pattern is that makes it hard to follow the logic sometimes, it could  make sense in this case but, because the code is simple enough, I don't mind the approach you have used. We might refactor the code later if needed.

File: infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/protobuf/schema/SchemaValidator.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+		}
+		for ( TableDefinition tableDefinition : owner.definitionsByTableName.values() ) {
+			if ( !tableDefinition.isDescribedIn( fileDescriptor ) ) {
+				throw LOG.providedSchemaHasAnInvalidCacheDefinition( tableDefinition.getTableName() );
+			}
+		}
+	}
+
+	private FileDescriptor parseSchema() {
+		// this name is used only for local parsing, it could be any value
```
[the remark this answers is not in the document]
Comment (fax4ever) [reply infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/protobuf/schema/SchemaValidator.java]: Protostream parsing API requires a file name as metadata of the Protobuf schema. This name may have no connection with the name that we use to register the schema to the server, it is used just for the parsing process.

[the remark this answers is not in the document]
Comment (anistor) [reply infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/protobuf/schema/SchemaValidator.java]: The name should be one and the same, or the eventual parsing error messages will stop being useful.

[the remark this answers is not in the document]
Comment (fax4ever) [reply infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/protobuf/schema/SchemaValidator.java]: @anistor thank you! you mean in case of parsing error, if I use a different name I will display the wrong name to the message log?

[the remark this answers is not in the document]
Comment (fax4ever) [reply infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/protobuf/schema/SchemaValidator.java]: I'm going to use the same name

[the remark this answers is not in the document]
Comment (DavideD) [reply infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/protobuf/schema/SchemaValidator.java]: Mmmh ... `any-name.proto` doesn't seem a good name ... if it appears in the error message

[the remark this answers is not in the document]
Comment (anistor) [reply infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/protobuf/schema/SchemaValidator.java]: Please use the real name

File: infinispan-remote/src/test/java/org/hibernate/ogm/datastore/infinispanremote/test/protobuf/ProtoSchemaOverrideResourceTest.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+			assertThat( schemaOverrideResource.getFile() ).endsWith( resource );
+
+			String expectedProtoSchema = TestHelper.readResource( schemaOverrideResource );
+
+			RemoteCache<Object, Object> cache = provider.getCache( ProtobufMetadataManagerConstants.PROTOBUF_METADATA_CACHE_NAME );
+			String protoSchema = (String) cache.get( provider.getSchemaFileName() );
+			assertThat( protoSchema ).isEqualTo( expectedProtoSchema );
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (fax4ever) [reply infinispan-remote/src/test/java/org/hibernate/ogm/datastore/infinispanremote/test/protobuf/ProtoSchemaOverrideResourceTest.java]: @DavideD I think there are no API for that

File: core/src/test/java/org/hibernate/ogm/utils/TestHelper.java [hunk 5]
```diff
@@ -428,4 +431,14 @@ public static void enableCountersForInfinispan(@SuppressWarnings("rawtypes") Map
 		cfg.put( "hibernate.ogm.infinispan.configuration_resource_name", "infinispan-dist.xml" );
 	}
 
+	public static String readResource(URL resource) {
+		try ( InputStream is = resource.openStream() ) {
+			java.util.Scanner s = new java.util.Scanner( is ).useDelimiter( "\\A" );
```
[the remark this answers is not in the document]
Comment (fax4ever) [reply core/src/test/java/org/hibernate/ogm/utils/TestHelper.java]: of course

File: documentation/manual/src/main/asciidoc/modules/infinispan.asciidoc [hunk 6]
```diff
@@ -2409,6 +2409,12 @@ hibernate.ogm.infinispan_remote.schema_file_name::
 Defines the file name of the generated Protobuf schema. Defualts to 'Hibernate_OGM_Generated_schema.proto'.
```
Comment (gsmet) [inline documentation/manual/src/main/asciidoc/modules/infinispan.asciidoc]: s/Defualts/Defaults/

File: documentation/manual/src/main/asciidoc/modules/infinispan.asciidoc [hunk 7]
```diff
@@ -2409,6 +2409,12 @@ hibernate.ogm.infinispan_remote.schema_file_name::
...
 The file name must have a valid __*.proto__ extension.
 
+hibernate.ogm.infinispan_remote.schema_override_resource::
```
[the remark this answers is not in the document]
Comment (gsmet) [reply documentation/manual/src/main/asciidoc/modules/infinispan.asciidoc]: Let's use a Unix path instead of a Windows one :).

File: core/src/test/java/org/hibernate/ogm/utils/TestHelper.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (fax4ever) [reply core/src/test/java/org/hibernate/ogm/utils/TestHelper.java]: I think that scanner method close just closes the input stream

File: core/src/test/java/org/hibernate/ogm/utils/TestHelper.java [hunk 8]
```diff
@@ -428,4 +431,14 @@ public static void enableCountersForInfinispan(@SuppressWarnings("rawtypes") Map
 		cfg.put( "hibernate.ogm.infinispan.configuration_resource_name", "infinispan-dist.xml" );
 	}
 
+	public static String readResource(URL resource) {
```
[the remark this answers is not in the document]
Comment (fax4ever) [reply core/src/test/java/org/hibernate/ogm/utils/TestHelper.java]: good idea

File: documentation/manual/src/main/asciidoc/modules/infinispan.asciidoc (hunk 7 above)
[the remark this answers is not in the document]
Comment (DavideD) [reply documentation/manual/src/main/asciidoc/modules/infinispan.asciidoc]: I don't mind, but I'm afraid that 90% of the developer works on Windows

[the remark this answers is not in the document]
Comment (gsmet) [reply documentation/manual/src/main/asciidoc/modules/infinispan.asciidoc]: At some point, they will deploy it on Linux, won't they? :)

I just think it's more appropriate to use a Unix path but, well, I won't fight for it if you think a Windows one is better. Note that there are also a lot of developers on MacOS these days and the Unix path works well for them too.

[the remark this answers is not in the document]
Comment (DavideD) [reply documentation/manual/src/main/asciidoc/modules/infinispan.asciidoc]: I like your world, let's live there :)

Anyway Unix works for me, I don't really mind

File: core/src/test/java/org/hibernate/ogm/utils/TestHelper.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (DavideD) [reply core/src/test/java/org/hibernate/ogm/utils/TestHelper.java]: Fair enough. Why the package declaration BTW?

File: documentation/manual/src/main/asciidoc/modules/infinispan.asciidoc (hunk 7 above)
[the remark this answers is not in the document]
Comment (Sanne) [reply documentation/manual/src/main/asciidoc/modules/infinispan.asciidoc]: I would like to see some Windows examples too... but then make sure to test them ;-)

[the remark this answers is not in the document]
Comment (fax4ever) [reply documentation/manual/src/main/asciidoc/modules/infinispan.asciidoc]: According to the java doc of ConfigurationPropertyReader class:
*in case of URLs this string can either represent a class path element, an URL or a file system path*
so I think that we can rely on helper, used for all URL extractions from any configuration.

[the remark this answers is not in the document]
Comment (fax4ever) [reply documentation/manual/src/main/asciidoc/modules/infinispan.asciidoc]: What do you think about something like:
> The value of the property is of string type and it can either represent a *class path element*, an *URL* or a *file system path*.

[the remark this answers is not in the document]
Comment (DavideD) [reply documentation/manual/src/main/asciidoc/modules/infinispan.asciidoc]: `The value of the property can either be a class path (example) , an URL (example) 
 or a file system path (example) .`

File: documentation/manual/src/main/asciidoc/modules/infinispan.asciidoc [hunk 9]
```diff
@@ -2409,6 +2409,12 @@ hibernate.ogm.infinispan_remote.schema_file_name::
...
+It is possible to override the generated Protobuf schema, providing a user defined Protobuf schema in a resource file.
+This will __not affect how entities are encoded__, so the alternative schema **must be compatible** with what would have been the generated one.
+In a nutshell the user can change only the comment parts of the Protobuf schema.
+Anyway this can be very useful for defining **server side indexes** on caches, thus used for JPQL or Ickle queries to data store.
+
```
[the remark this answers is not in the document]
Comment (gsmet) [reply documentation/manual/src/main/asciidoc/modules/infinispan.asciidoc]: s/defined/define/

**Comment by fax4ever** [comment]
Thank you for comments and suggestions. I did some changes. That anything else we can improve or fix?

File: core/src/main/java/org/hibernate/ogm/util/impl/ResourceHelper.java [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+ */
+package org.hibernate.ogm.util.impl;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+
+public class ResourceHelper {
+
+	private static final String FILE_DELIMITER_PATTERN = "\\A";
```
Comment (DavideD) [inline core/src/main/java/org/hibernate/ogm/util/impl/ResourceHelper.java]: Is this a "file delimiter" or a "line delimiter"? Does it read multiple files?

File: infinispan-remote/src/test/java/org/hibernate/ogm/datastore/infinispanremote/test/protobuf/ProtoSchemaOverrideResourceTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (DavideD) [reply infinispan-remote/src/test/java/org/hibernate/ogm/datastore/infinispanremote/test/protobuf/ProtoSchemaOverrideResourceTest.java]: Shouldn't you be able to add a native full text condition on a field that it is indexed? Otherwise it shouldn't work. Am I wrong?

[the remark this answers is not in the document]
Comment (DavideD) [reply infinispan-remote/src/test/java/org/hibernate/ogm/datastore/infinispanremote/test/protobuf/ProtoSchemaOverrideResourceTest.java]: It doesn't matter, we can do it later

File: core/src/main/java/org/hibernate/ogm/util/impl/ResourceHelper.java (hunk 10 above)
Comment (DavideD) [reply core/src/main/java/org/hibernate/ogm/util/impl/ResourceHelper.java]: Are you going to change it? Otherwise I can do it before merging the PR. Let me know

Comment (fax4ever) [reply core/src/main/java/org/hibernate/ogm/util/impl/ResourceHelper.java]: not yet

Comment (DavideD) [reply core/src/main/java/org/hibernate/ogm/util/impl/ResourceHelper.java]: Don't worry, I will do that

Comment (fax4ever) [reply core/src/main/java/org/hibernate/ogm/util/impl/ResourceHelper.java]: thank you

**Comment by fax4ever** [comment]
@DavideD I'm going to split the commit

**Comment by DavideD** [comment]
> @DavideD I'm going to split the commit

Thanks

**Comment by fax4ever** [comment]
@DavideD I split the old commit and pushed with force the new ones

**Comment by DavideD** [comment]
Merged https://github.com/hibernate/hibernate-ogm/pull/1033

I made a mistake and merged the previous version in  a single commit.
